File: core/sources/reference-list.ts
import { IndexedSourceBase, type Metadata } from "../types/source";

/**
 * Source that resolves a list on demand.
 */
export class ReferenceListSource extends IndexedSourceBase {
  readonly kind = "reference-list";

  /**
   * @param id Unique source ID
   * @param metadata Generated metadata for the source
   * @param getContent Resolver for list items
   */
  constructor(
    id: string,
    metadata: Metadata | undefined,
    /** Returns the resolved list items. */
    public getContent: () =>
      | Promise<AsyncIterable<string>>
      | AsyncIterable<string>,
  ) {
    super(id, metadata);
    this.getContent = getContent;
  }

  get isIterable(): true {
    return true;
  }

  async buildIndexInput(): Promise<string> {
    const content = await this.getContent();
    const items: string[] = [];
    for await (const item of content) items.push(item);
    return items.join("\n");
  }

  async buildDeepJudgeInput(): Promise<string> {
    const content = await this.getContent();
    const items: string[] = [];
    for await (const item of content) {
      items.push(item);
      if (items.length >= 10) break;
    }
    return items.join("\n");
  }
}

File: python/src/contextrie/sources/reference_list.py
from __future__ import annotations

from collections.abc import AsyncIterable, Awaitable, Callable

from .._runtime import collect_async_iterable, maybe_await
from ..types import IndexedSourceBase, Metadata


class ReferenceListSource(IndexedSourceBase):
    """Source that resolves a list on demand."""

    kind = "reference-list"

    def __init__(
        self,
        id: str,
        metadata: Metadata | None,
        get_content: Callable[[], AsyncIterable[str] | Awaitable[AsyncIterable[str]]],
        path: str | None = None,
    ) -> None:
        super().__init__(id=id, metadata=metadata, path=path)
        self._get_content = get_content

    @property
    def is_iterable(self) -> bool:
        return True

    def get_content(self) -> AsyncIterable[str] | Awaitable[AsyncIterable[str]]:
        return self._get_content()

    async def build_index_input(self) -> str:
        content = await maybe_await(self._get_content())
        items = await collect_async_iterable(content)
        return "\n".join(items)

    async def build_deep_judge_input(self) -> str:
        content = await maybe_await(self._get_content())
        items = await collect_async_iterable(content, limit=10)
        return "\n".join(items)
